AIP-86 - Add doc pages for the new Deadline Alerts feature (#53727)
mistercrunch/Airflow: airflow-core/docs/howto/deadline-alerts.rst

 .. Licensed to the Apache Software Foundation (ASF) under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

 ..   http://www.apache.org/licenses/LICENSE-2.0

 .. Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

Deadline Alerts
===============

The :class:`~airflow.sdk.definitions.deadline.DeadlineAlert` feature is the next evolution of

> the old SLA.  Deadline Alerts allow you to set time thresholds for your DAG runs and automatically

Reviewer: I think it has been brought up we should not compare Airflow version in docs, I am just wondering if mentioning "the old SLA" goes against that.

Reviewer 2: Currently, the [Tasks docs](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/tasks.html#slas) metion:

```The SLA feature from Airflow 2 has been removed in 3.0 and will be replaced with a new implementation in Airflow 3.1```

So we could update the above to something like 

```The SLA feature from Airflow 2 has been removed in 3.0 and was replaced in Airflow 3.1 with Deadlines Alerts (+link)```

And delete the comparison in this section (```The :class:`~airflow.sdk.definitions.deadline.DeadlineAlert` feature is the next evolution of
the old SLA.```).

What do you think?

Author: I think it's a good solution.  I'll make it happen.  If someone has a better solution, we can always change it again later.

     Start time                          Trigger point

Here's an example DAG implementation. If the DAG has not finished 15 minutes after it was queued, send an email:

.. code-block:: python

    from datetime import datetime, timedelta
    from airflow import DAG
    from airflow.sdk.definitions.deadline import DeadlineAlert, DeadlineReference
    from airflow.providers.smtp.notifications.smtp import SmtpNotifier
    from airflow.providers.standard.operators.empty import EmptyOperator

    with DAG(
        dag_id="deadline_alert_example",
        deadline=DeadlineAlert(
            reference=DeadlineReference.DAGRUN_QUEUED_AT,
            interval=timedelta(minutes=15),
            callback=SmtpNotifier(
                to="team@example.com",
                subject="[Alert] DAG {{ dag.dag_id }} exceeded time threshold",
                html_content="The DAG has been running for more than 15 minutes since being queued.",
            ),
        ),
    ):
        EmptyOperator(task_id="example_task")

The timeline for this example would look like this:

::

    |------|-----------|---------|-----------|--------|
        Scheduled    Queued    Started    Deadline

>          00:00       00:03      00:05      00:18

Reviewer: It is 15 minutes right? Other than that, I love this section, it makes it super easy to understand the concept of deadlines!

```suggestion
         00:00       00:03      00:05      00:15
```

Author: Yup, but it's 15 minutes after the QUEUED_AT in this example, so 0:18 is right.  If you used the LOGICAL_DATE reference i would be at 0:15

Reviewer: Oh! My bad, I was not attentive enough

Using Built-in References
-------------------------

Airflow provides several built-in reference points that you can use with DeadlineAlert:

``DeadlineReference.DAGRUN_QUEUED_AT``
    Measures time from when the DagRun was queued. Useful for monitoring resource constraints.

``DeadlineReference.DAGRUN_LOGICAL_DATE``
    References when the DAG run was scheduled to start. For example, setting an interval of
    ``timedelta(minutes=15)`` would trigger the alert if the DAG hasn't completed 15 minutes
    after it was scheduled to start, regardless of when (or if) it actually began executing.
    Useful for ensuring scheduled DAGs complete before their next scheduled run.

``DeadlineReference.FIXED_DATETIME``
    Specifies a fixed point in time. Useful when DAGs must complete by a specific time.

Here's an example using a fixed datetime:

.. code-block:: python

    tomorrow_at_ten = datetime.combine(datetime.now().date() + timedelta(days=1), time(10, 0))

    with DAG(
        dag_id="fixed_deadline_alert",
        deadline=DeadlineAlert(
            reference=DeadlineReference.FIXED_DATETIME(tomorrow_at_ten),

>             interval=timedelta(minutes=-30),  # Alert 30 minutes before the deadline

Reviewer: ```suggestion
            interval=timedelta(minutes=-30),  # Alert 30 minutes before the reference
```

Author: That's fair, I'll make the change.

            callback=SmtpNotifier(
                to="team@example.com",
                subject="Report will be late",
                html_content="The report will not be ready 30 minutes before the deadline.",
            ),
        ),
    ):
        EmptyOperator(task_id="example_task")

The timeline for this example would look like this:

::

    |------|----------|---------|----------|--------|
        Deadline    Queued    Start    Reference

>          08:00      09:15     09:17     10:00

Reviewer: Is this timeline right? It does not seem to match the description above

Author: Oh man, you are right, I changed the example and never updated the timeline!  Good catch, thanks!  It should be 

```
    |------|------------|----------|--------------|--------|
        Queued    Start   Deadline   Reference
          09:15      09:17       09:30       10:00
```         

Given that this one is "backwards" (negative interval)  and the deadline is a time before the reference, do you think it is any easier to understand with an arrow or some indicator like this:

```
    |------|------------|----------|-----------------|--------|
        Queued    Start    Deadline       Reference
          09:15      09:17       09:30   <==   10:00
```

Reviewer: I understood without it so I am not sure this is needed, but it does not harm having it either. I'll let you decide :)

Author: Alright, for mow I'll just fix the layout and leave off the arrow.  Thanks.

Using Callbacks
---------------

When a deadline is exceeded, the callback is executed. You can use any async :doc:`Notifier </howto/notifications>`
or create a custom callback function.

Using Built-in Notifiers
^^^^^^^^^^^^^^^^^^^^^^^^

Here's an example using the Slack notifier if the DagRun has not finished within 30 minutes of it being queued:

.. code-block:: python

    with DAG(
        dag_id="slack_deadline_alert",
        deadline=DeadlineAlert(
            reference=DeadlineReference.DAGRUN_QUEUED_AT,

>             interval=timedelta(minutes=15),

Reviewer: ```suggestion
            interval=timedelta(minutes=30),
```

Author: Thanks so much for catching these.  I went through a bunch of revisions and clearly missed some edits.  A bit embarrassing....

Reviewer: No worries :) It is always harder to catch these when you're head down :)

    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

 ..   http://www.apache.org/licenses/LICENSE-2.0

 .. Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

Deadline Alerts
===============

The :class:`~airflow.sdk.definitions.deadline.DeadlineAlert` feature is the next evolution of
the old SLA.  Deadline Alerts allow you to set time thresholds for your DAG runs and automatically
respond when those thresholds are exceeded. You can set up Deadline Alerts by choosing a built-in
reference point, setting an interval, and defining a response using either Airflow's Notifiers or
a custom callback function.

Creating a Deadline Alert
-------------------------

> To create a Deadline Alert, there are three three components you must specify and one optional one:

Reviewer 3: ```suggestion
To create a Deadline Alert, there are three components you must specify and one optional one:
```

Reviewer 3: Alternative phrasing: "three mandatory and one optional parameter"

Author: Thanks, I'll get it in the upcoming revision.  I've changed it to "three required...." which I think is what my brain meant to type but my fingers didn't get the memo.

``DeadlineReference.FIXED_DATETIME``
    Specifies a fixed point in time. Useful when DAGs must complete by a specific time.

Here's an example using a fixed datetime:

.. code-block:: python

    tomorrow_at_ten = datetime.combine(datetime.now().date() + timedelta(days=1), time(10, 0))

    with DAG(
        dag_id="fixed_deadline_alert",
        deadline=DeadlineAlert(
            reference=DeadlineReference.FIXED_DATETIME(tomorrow_at_ten),
            interval=timedelta(minutes=-30),  # Alert 30 minutes before the reference.
            callback=SmtpNotifier(
                to="team@example.com",
                subject="Report will be late",
                html_content="The report will not be ready 30 minutes before the deadline.",
            ),
        ),
    ):
        EmptyOperator(task_id="example_task")

The timeline for this example would look like this:

::

    |------|----------|---------|------------|--------|
         Queued     Start    Deadline    Reference
         09:15      09:17     09:30       10:00

.. note::
    Note that since the interval is a negative value, the deadline is before the reference in this case.

Using Callbacks
---------------

> When a deadline is exceeded, the callback is executed. You can use any async :doc:`Notifier </howto/notifications>`

Reviewer 4: ```suggestion
When a deadline is exceeded, the callback is executed. You can use any :doc:`Notifier </howto/notifications>`
```
Using sync or async is the goal right? Did you specifically mention async here because the sync path is still in progress? If so I'd just do something like:

```suggestion
When a deadline is exceeded, the callback is executed. You can use any :doc:`Notifier </howto/notifications>` (Note: that synchronous callback support is in development)
```

Author: I actually have that adjusted in a current revision.  Yes, that was my reasoning, but we're going to fast-track the sync work so I am tweaking it now.

Author: What I have there right now is 

===

When a deadline is exceeded, the callback is executed. You can use an existing :doc:`Notifier </howto/notifications>` or create a custom callback function.

===

I'm just waiting for build-docs and static checks to run and I'll be pushing that change and a rework of the custom callbacks section to add more non-async context.

or create a custom callback function.

Using Built-in Notifiers
^^^^^^^^^^^^^^^^^^^^^^^^

Here's an example using the Slack notifier if the DagRun has not finished within 30 minutes of it being queued:

.. code-block:: python

    with DAG(
        dag_id="slack_deadline_alert",
        deadline=DeadlineAlert(
            reference=DeadlineReference.DAGRUN_QUEUED_AT,
            interval=timedelta(minutes=30),
            callback=SlackNotifier(
                slack_conn_id="slack_default",
                channel="#alerts",
                text="DAG {{ dag.dag_id }} has been running for more than 30 minutes since being queued.",
                username="Airflow Alerts",
            ),
        ),
    ):
        EmptyOperator(task_id="example_task")

Creating Custom Callbacks
^^^^^^^^^^^^^^^^^^^^^^^^^

You can create custom callbacks for more complex handling. The ``callback_kwargs`` specified in

> the ``DeadlineAlert`` are passed to the callback function.

Reviewer 4: This and the code snippets need updating after https://github.com/apache/airflow/pull/53951#pullrequestreview-3080641355 right?

Author: That is possible.   I'll keep an eye on the in-flight PRs and adjust accordingly.

Author: Updated in https://github.com/apache/airflow/pull/53727/commits/28d88e269bbb2123557f75f4716550c95981eb8a

        # Additional custom handling here

    # Place this in a dag file
    from datetime import timedelta

    from deadline_callbacks import custom_callback

    from airflow import DAG
    from airflow.providers.standard.operators.empty import EmptyOperator
    from airflow.sdk.definitions.deadline import DeadlineAlert, DeadlineReference

    with DAG(
        dag_id="custom_deadline_alert",
        deadline=DeadlineAlert(
            reference=DeadlineReference.DAGRUN_QUEUED_AT,
            interval=timedelta(minutes=15),
            callback=custom_callback,
            callback_kwargs={"alert_type": "time_exceeded", "dag_id": "custom_deadline_alert"},
        ),
    ):
        EmptyOperator(task_id="example_task")

.. note::
    Regarding Deadline callbacks:

    * Async callbacks are recommended as they will be executed by the Triggerer.
    * Users must ensure any async callback is importable by the Triggerer.
    * One easy way to do this is to place the callback as a top-level method in a new file in the plugins folder.
    * The Triggerer may need to be restarted when a callback is added or changed in order to reload the files.

> Deadline Calculation

Reviewer 4: This block is really nice to hit home the notion of how this all works. Maybe even pull it closer to the top.

 .. Licensed to the Apache Software Foundation (ASF) under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

 ..   http://www.apache.org/licenses/LICENSE-2.0

 .. Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

Deadline Alerts
===============

The :class:`~airflow.sdk.definitions.deadline.DeadlineAlert` feature is the next evolution of

> the old SLA.  Deadline Alerts allow you to set time thresholds for your DAG runs and automatically

Reviewer 5: ```suggestion
the old SLA.  Deadline Alerts allow you to set time thresholds for your Dag runs and automatically
```

respond when those thresholds are exceeded. You can set up Deadline Alerts by choosing a built-in
reference point, setting an interval, and defining a response using either Airflow's Notifiers or
a custom callback function.

Creating a Deadline Alert
-------------------------

> Creating a Deadline Alert requires three mandatory and one optional parameter:

Reviewer 5: ```suggestion
Creating a Deadline Alert requires 3 mandatory and 1 optional parameter:
```

I feel it might be more readable this way 🤔

Author: This line is no longer really relevant since https://github.com/apache/airflow/pull/53951/files removed the optional one, I'll be rephrasing it.

* Reference: When to start counting from
* Interval: How far before or after the reference point to trigger the alert
* Callback: What to do when the deadline is exceeded

> * Callback Kwargs:  Optional values to pass to the Callback when it is run

Reviewer 5: ```suggestion
* Callback Kwargs: Optional values to pass to the Callback when it is run
```

> * Callback Kwargs:  Optional values to pass to the Callback when it is run

Reviewer 5: ```suggestion
* ``reference``: When to start counting from
* ``interval``: How far before or after the reference point to trigger the alert
* ``callback``: What to do when the deadline is exceeded
```

looks more like 3 mandatory args but we can modify callback

Author: Good catch, it needs to be updated; there are now three mandatory and no optional after @ramitkataria 's recent PR merged.

Here is how Deadlines are calculated:

::

    [Reference] ------ [Interval] ------> [Deadline]
        ^                                     ^
        |                                     |
     Start time                          Trigger point

> Below is an example DAG implementation. If the DAG has not finished 15 minutes after it was queued, send a Slack message:

Reviewer 5: ```suggestion
Below is an example Dag implementation. If the Dag has not finished 15 minutes after it was queued, send a Slack message:
```

Airflow provides several built-in reference points that you can use with DeadlineAlert:

``DeadlineReference.DAGRUN_QUEUED_AT``
    Measures time from when the DagRun was queued. Useful for monitoring resource constraints.

``DeadlineReference.DAGRUN_LOGICAL_DATE``
    References when the DAG run was scheduled to start. For example, setting an interval of
    ``timedelta(minutes=15)`` would trigger the alert if the DAG hasn't completed 15 minutes
    after it was scheduled to start, regardless of when (or if) it actually began executing.
    Useful for ensuring scheduled DAGs complete before their next scheduled run.

``DeadlineReference.FIXED_DATETIME``
    Specifies a fixed point in time. Useful when DAGs must complete by a specific time.

Here's an example using a fixed datetime:

.. code-block:: python

    tomorrow_at_ten = datetime.combine(datetime.now().date() + timedelta(days=1), time(10, 0))

    with DAG(
        dag_id="fixed_deadline_alert",
        deadline=DeadlineAlert(
            reference=DeadlineReference.FIXED_DATETIME(tomorrow_at_ten),
            interval=timedelta(minutes=-30),  # Alert 30 minutes before the reference.
            callback=AsyncCallback(

>                 SlackWebhookNotifier,

Reviewer 6: This should be awaitable for `AsyncCallback ` , SlackWebhookNotifier doesn't have any awaitbale's. IMHO its better to keep working example ? WDYT?

Reviewer 7: Right now in `main`, there's no awaitable Notifiers but once #53831 is merged `SlackWebhookNotifier` will be awaitable and we'll make a PR for a couple more awaitable notifiers as well. So I think the plan is to merge that PR first (some unit tests still require some work) and then merge this one so it will be a working example

Author: Hey, sorry, I've been on a camping trip.   First, thanks for the review and I appreciate the catch!  Second, what Ramit said :P   At this point, this is kind of a "living document" since it's documenting a feature that is almost but not quite done.  See the comments below about callback_kwargs as another example.  Tha Notifier will be awaitable before this is merged though.

                kwargs={
                    "slack_conn_id": "slack_default",
                    "channel": "#alerts",
                    "text": "DAG 'slack_deadline_alert' still running after 30 minutes.",
                    "username": "Airflow Alerts",
                },
            ),
        ),
    ):
        EmptyOperator(task_id="example_task")

The timeline for this example would look like this:

::

    |------|----------|---------|------------|--------|
         Queued     Start    Deadline    Reference
         09:15      09:17     09:30       10:00

.. note::
    Note that since the interval is a negative value, the deadline is before the reference in this case.

Using Callbacks
---------------

When a deadline is exceeded, the callback is executed. You can use an existing :doc:`Notifier </howto/notifications>`
or create a custom callback function.  A callback must be either an :class:`~airflow.sdk.definitions.deadline.AsyncCallback`
or a :class:`~airflow.sdk.definitions.deadline.SyncCallback`.

Using Built-in Notifiers
^^^^^^^^^^^^^^^^^^^^^^^^

> Here's an example using the Slack Notifier if the DagRun has not finished within 30 minutes of it being queued:

Reviewer 5: ```suggestion
Here's an example using the Slack Notifier if the Dag run has not finished within 30 minutes of it being queued:
```

.. code-block:: python

    with DAG(
        dag_id="slack_deadline_alert",
        deadline=DeadlineAlert(
            reference=DeadlineReference.DAGRUN_QUEUED_AT,
            interval=timedelta(minutes=30),
            callback=AsyncCallback(

>                 SlackNotifier,

Reviewer 6: Same as above comment.

Reviewer 7: I think we should use `SlackWebhookNotifier` here as well instead of `SlackNotifier` because that's the one we're merging first and will be a working example when this PR is merged

Author: Going to unresvolve this one for a bit so i remember to make Ramit's change.  Thanks for catching it!

                    "text": "Dag 'slack_deadline_alert' still running after 30 minutes.",
                    "username": "Airflow Alerts",
                },
            ),
        ),
    ):
        EmptyOperator(task_id="example_task")

Creating Custom Callbacks
^^^^^^^^^^^^^^^^^^^^^^^^^

You can create custom callbacks for more complex handling. The ``kwargs`` specified in the ``Callback``

> are passed to the callback function, if any are provided.  **Synchronous callbacks** (standard python

Reviewer 5: ```suggestion
are passed to the callback function, if any are provided. **Synchronous callbacks** (standard python
```

Author: Hey, stop picking on my age!  Double-spacer for life!  ((I'll change them))

> You can create custom callbacks for more complex handling. The ``callback_kwargs`` specified in

Reviewer 5: This `callback_kwargs` is a bit confusing to me as it's not the actual parameter name.

Author: It was at the time :P   https://github.com/apache/airflow/pull/53951 removed it, I need to update these docs to match now.  Thanks for pointing it out.

> methods) can be defined in the dag bundle and are run in the Executor.  **Asynchronous callbacks** must

Reviewer 5: ```suggestion
methods) can be defined in the dag bundle and are run in the Executor. **Asynchronous callbacks** must
```

A **custom synchronous callback** might look like this:

.. code-block:: python

    from datetime import timedelta

    from airflow import DAG
    from airflow.providers.standard.operators.empty import EmptyOperator
    from airflow.sdk.definitions.deadline import DeadlineAlert, DeadlineReference, SyncCallback

    def custom_synchronous_callback(**kwargs):
        """Handle deadline violation with custom logic."""
        print(f"Deadline exceeded for DAG {kwargs.get("dag_id")}!")
        print(f"Alert type: {kwargs.get("alert_type")}")
        # Additional custom handling here

    with DAG(
        dag_id="custom_deadline_alert",
        deadline=DeadlineAlert(
            reference=DeadlineReference.DAGRUN_QUEUED_AT,
            interval=timedelta(minutes=15),
            callback=SyncCallback(
                custom_synchronous_callback,
                kwargs={"alert_type": "time_exceeded", "dag_id": "custom_deadline_alert"},
            ),
        ),
    ):
        EmptyOperator(task_id="example_task")

A **custom asynchronous callback** is only slightly more work.  Note in the following example that

> the custom callback code is placed in a separate file, and must be imported in the DAG file.

Reviewer 5: ```suggestion
the custom callback code is placed in a separate file, and must be imported in the Dag file.
```

.. code-block:: python

    # Place this method in `/files/plugins/deadline_callbacks.py`
    async def custom_async_callback(**kwargs):
        """Handle deadline violation with custom logic."""
        print(f"Deadline exceeded for DAG {kwargs.get("dag_id")}!")
        print(f"Alert type: {kwargs.get("alert_type")}")
        # Additional custom handling here

>     # Place this in a dag file

Reviewer 5: ```suggestion

.. code-block:: python

    # Place this in a dag file
```

maybe making them 2 blocks?

Author: Sure, I'll make the change in the next revision, thanks.

    from datetime import timedelta

    from deadline_callbacks import custom_async_callback

    from airflow import DAG
    from airflow.providers.standard.operators.empty import EmptyOperator
    from airflow.sdk.definitions.deadline import AsyncCallback, DeadlineAlert, DeadlineReference

    with DAG(
        dag_id="custom_deadline_alert",
        deadline=DeadlineAlert(
            reference=DeadlineReference.DAGRUN_QUEUED_AT,
            interval=timedelta(minutes=15),
            callback=AsyncCallback(
                custom_async_callback,
                kwargs={"alert_type": "time_exceeded", "dag_id": "custom_deadline_alert"},
            ),
        ),
    ):
        EmptyOperator(task_id="example_task")

Deadline Calculation
^^^^^^^^^^^^^^^^^^^^

A deadline's trigger time is calculated by adding the ``interval`` to the datetime returned by
the ``reference``. For ``FIXED_DATETIME`` references, negative intervals can be particularly
useful to trigger the callback *before* the reference time.

For example:

.. code-block:: python

>     next_meeting = datetime(2025, 06, 26, 9, 30)

Reviewer 5: ```suggestion
    next_meeting = datetime(2025, 6, 26, 9, 30)
```

this is a syntax error

Author: Nice catch, thanks.  I saw that pre-commits ran mypy on the code snippets and had assumed it would have caught that.  Cheers!

    DeadlineAlert(
        reference=DeadlineReference.FIXED_DATETIME(next_meeting),
        interval=timedelta(hours=-2),
        callback=notify_team,
    )

This will trigger the alert 2 hours before the next meeting starts.

For ``DAGRUN_LOGICAL_DATE``, the interval is typically positive, setting a deadline relative

> to when the DAG was scheduled to run. Here's an example:

Reviewer 5: ```suggestion
to when the Dag was scheduled to run. Here's an example:
```

.. code-block:: python

    DeadlineAlert(
        reference=DeadlineReference.DAGRUN_LOGICAL_DATE,
        interval=timedelta(hours=1),
        callback=notify_team,
    )

> In this case, if a DAG is scheduled to run daily at midnight, the deadline would be triggered

Reviewer 5: ```suggestion
In this case, if a Dag is scheduled to run daily at midnight, the deadline would be triggered
```

> if the DAG hasn't completed by 1:00 AM. This is useful for ensuring that scheduled jobs complete

Reviewer 5: ```suggestion
if the Dag hasn't completed by 1:00 AM. This is useful for ensuring that scheduled jobs complete
```
